vividus-engine/src/main/java/org/vividus/steps/VariableResolver.java

```diff
@@ -156,13 +160,29 @@ private Object getVariableValue(String variableKey)
         Object variable = variableContext.getVariable(variableKey);
         if (variable == null)
         {
-            variable = execute(() -> Optional.ofNullable(dynamicVariables.get(variableKey))
-                               .map(DynamicVariable::getValue)
-                               .orElse(null), null);
+            variable = execute(() -> calculateDynamicVariableValue(variableKey), null);
         }
         return variable;
     }
 
+    private String calculateDynamicVariableValue(String variableKey)
+    {
+        DynamicVariable dynamicVariable = dynamicVariables.get(variableKey);
+        if (dynamicVariable != null)
+        {
+            Optional<String> value = dynamicVariable.calculateValue();
+            if (value.isPresent())
+            {
+                return value.get();
+            }
+            LOGGER.atError()
+                    .addArgument(variableKey)
+                    .addArgument(dynamicVariable::getMissingValueErrorMessage)
+                    .log("Unable to resolve dynamic variable ${{}}: {}");
```

Comment:
it makes sense to log that there is no such dynamic variable

Reply:
we don't know which variable is looked up by user: either regular one or dynamic variable